### PR TITLE
Padding zero run

### PR DESCRIPTION
**Pull request by dkg**

### CODE REVIEW

File: draft-ietf-tls-tls13.md [hunk 1]
```diff
@@ -991,9 +992,10 @@ compatibility, the record layer version identifies as simply TLS 1.0.
 Endpoints supporting other versions negotiate the version to use
 by following the procedure and requirements in {{backward-compatibility}}.
 
-Implementations MUST NOT send zero-length fragments of Handshake or Alert
-types. Zero-length fragments of Application data MAY
-be sent as they are potentially useful as a traffic analysis countermeasure.
+Implementations MUST NOT send zero-length fragments of Handshake or
+Alert types, even if those fragments contain padding. Zero-length
+fragments of Application data MAY be sent as they are potentially
```
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: Why is Application capitalized here?

File: draft-ietf-tls-tls13.md [hunk 2]
```diff
@@ -1032,8 +1040,22 @@ length
...
   this length MUST generate a fatal "record_overflow" alert.
 
+fragment.content
+: The cleartext of TLSPlaintext.fragment.
+
+fragment.type
+: The actual content type of the record.
+
+fragment.zeros
+: An arbitrary-length run of null bytes (uint8 with value 0) may
```
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: An arbitrary-length run of zero-valued octets, ...

File: draft-ietf-tls-tls13.md (hunk 1 above)
Comment (davegarrett) [inline draft-ietf-tls-tls13.md]: Looks to be a pre-existing typo. A fix to lowercase it here is probably a good idea.

File: draft-ietf-tls-tls13.md [hunk 3]
```diff
@@ -1084,20 +1108,71 @@ In order to decrypt and verify, the cipher takes as input the key, nonce, the
...
+
+### Record Padding
+
+All encrypted TLS records can be padded to inflate the size of the
+TLSCipherText.  This allows the sender to hide the size of the
+traffic from an observer.
+
+When generating a TLSCiphertext record, implementations MAY choose to
+pad.  An unpadded record is just a record with a padding length of
+zero.  Padding is a string of null bytes (uint8 with value 0) appended
```
Comment (martinthomson) [inline draft-ietf-tls-tls13.md]: zero-valued octets

**Comment by davegarrett** [comment]
This is a big improvement over the previous proposed methods, by the way.

@martinthomson: I suggest filing a followup PR for your comments, as it looks like this has already merged.

File: draft-ietf-tls-tls13.md [hunk 4]
```diff
@@ -1084,20 +1108,71 @@ In order to decrypt and verify, the cipher takes as input the key, nonce, the
...
+This permits generation of plausibly-sized cover traffic in contexts
+where the presence or absence of activity may be sensitive.
+Implementations MUST NOT send Handshake or Alert records that are all
+padding.
+
+The padding sent will be authenticated by the record protection
+mechanism.  Upon successful decryption of a TLSCiphertext.fragment,
+the receiving implementation scans the field from the end toward the
+beginning until it finds a non-zero uint8.  This non-zero octet is the
+content type of the message.
```
Comment (davegarrett) [inline draft-ietf-tls-tls13.md]: It should probably explicitly state where the content data is, too.

**Pull request merged**